troubleshooting (#80)
skupperproject/skupper-docs: modules/troubleshooting/pages/index.adoc

include::{mod-loc}attributes.adoc[]

//Category: skupper-troubleshooting
// Type: assembly

[id="troubleshooting"] 
= Troubleshooting a {service-network}

[role="system:abstract"]
Typically, you can create a {service-network} without referencing this troubleshooting guide.
However, this guide provides some tips for situations when the {service-network} does not perform as expected.

See xref:common-problems[] if you have encountered a specific issue using the `skupper` CLI.

A typical troubleshooting workflow is to check all the sites, and create debug tarballs.

// Type: procedure
[id="checking-sites"] 
== Checking sites

[role="system:abstract"]
Using the `skupper` command-line interface (CLI) provides a simple method to get started with troubleshooting {skupper-name}.

.Procedure

. Check the site status:
+
--
----
$ skupper status --namespace west

> Skupper is enabled for namespace "west" in interior mode. It is connected to 2 other sites. It has 1 exposed services.

Reviewer: ```suggestion
Skupper is enabled for namespace "west" in interior mode. It is connected to 2 other sites. It has 1 exposed service.
```
Gotta be consistent on on using numerals or words to depict numbers. Numerals are used in this sentence but "two" is used a few sentences later. See https://www.ibm.com/docs/en/ibm-style?topic=measurement-numerals-versus-words

Author: changed later instance 'two'->'2'
Cannot change this text as it's cli output

----

The output shows:

* A site exists in the specified namespace.
* A link exists to two other sites.

> * A service is exposed on the {service-network}, and is accessible from this namespace.

Reviewer: ```suggestion
* A service is exposed on the {service-network} and is accessible from this namespace.
```
Passive voice also.

Author: dunno how to unpassive this one :(

--

. Check the link status:
+
--
----
$ skupper link status --namespace east

Link link1 is active
----

> A link exists from the specified site to another site, meaning a token from another site was applied to the specified site.

Reviewer: ```suggestion
A link exists from the specified site to another site, meaning a token from another site applies to the specified site.
```
Passive voice.

Author: doesn't work, need past tense, and in this case we don't know who did it... so leaving as is

> NOTE: Running `skupper link status` on a connected site only produces output if a token was used to create a link.   

Reviewer: ```suggestion
NOTE: Running `skupper link status` on a connected site produces output only if a token was used to create a link.   
```
https://www.ibm.com/docs/en/ibm-style?topic=grammar-adverbs-only
Also, passive voice.

--

. Check the {service-network}:
+
--
----
$ skupper network status --namespace west

Sites:
├─ [local] 05f8c38 - west 
│  URL: 10.110.15.54
│  mode: interior
│  name: west
│  namespace: west
│  version: 1.0.2
│  ╰─ Services:
│     ╰─ name: backend
│        address: backend: 8080
│        protocol: tcp
╰─ [remote] 1537b82 - east 
   URL: 10.97.26.100
   name: east
   namespace: east
   sites linked to: 05f8c38-west
   version: 1.0.2
   ╰─ Services:
      ╰─ name: backend
         address: backend: 8080
         protocol: tcp
         ╰─ Targets:
            ├─ name: backend-77f8f45fc8-smckp
            ├─ name: backend-77f8f45fc8-gh6tp
            ╰─ name: backend-77f8f45fc8-m58tg
----

The output shows:

* There are two sites on the {service-network}, `east` and `west`.

> * Details for each site, for example the namespace names.

Reviewer: ```suggestion
* Details for each site. For example, the namespace names.
```

> * The original services that are exposed (Targets), in this case the three backend services exposed using the `tcp` protocol.

Reviewer: ```suggestion
* The original services that are exposed (Targets). In this case, the three backend services exposed using the `tcp` protocol.
```

> * The services available on the {service-network}, including port, for example, `backend:8080`. 

Reviewer: ```suggestion
* The services available on the {service-network}, including the port. For example, `backend:8080`. 
```

--

// Type: procedure
[id='creating-debug']
== Creating a skupper debug tarball

> The debug tarball contains all the logs from the {skupper-name} components, and provides detailed information to help debug issues.

Reviewer: ```suggestion
The debug tarball contains all the logs from the {skupper-name} components and provides detailed information to help debug issues.
```

. Create the debug tarball:
+
--
----
$  skupper debug dump my-site

Skupper dump details written to compressed archive:  my-site.tar.gz

----

--

// Type:procedure
[id='common-problems']
== Resolving common problems

The following issues and workarounds might help you debug simple scenarios when evaluating {skupper-name}.

*Cannot initialize skupper*

If the `skupper init` command fails, consider the following options:

* Check the load balancer.
+
--

> If you are evaluating {skupper-name} on minikube, a load balancer is not automatically created.

Reviewer: Passive voice.

> Use the following command to create a loadbalancer:

Reviewer: ```suggestion
Use the following command to create a load balancer:
```

----
$ minikube tunnel
----

For other Kubernetes flavors, see the documentation from your provider.

--

* Initialize without ingress.
+
--
This option will prevent other sites from linking to this site, but linking outwards is supported.

> Once a link is established, traffic can flow in either direction.

Reviewer: Passive voice.

Author: the future tense in the prev sentence is prob more egregious, and preventable ;)
